Approximate district heating COPs via Jensen et al. 2018

amosschle opened the pull request:
Closes #857.

with respect to central heating.

## Changes proposed in this Pull Request
- Introduce separate forward and return temperatures for district (central) heating.
- The rule `build_cop_profiles` builds separate heat pump COPs for decentral and central heating.
- The COP approximation for central heating is based on [Jensen et al. 2018](https://backend.orbit.dtu.dk/ws/portalfiles/portal/151965635/MAIN_Final.pdf).
- Additional config settings for the approximation are added, with default values partly based on [Pieper et al. 2019](https://www.sciencedirect.com/science/article/pii/S0360544220311014)

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Changed dependencies are added to `envs/environment.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv`.
- [x] A release note `doc/release_notes.rst` is added.

Jupyter notebook and config files are attached below (`.txt` must be changed accordingly).

## Impact on results
A quick test run (DE only, 2 nodes, 2050) suggests the resulting COPs after changing the approximation method **and** forward/return temperatures in district heating to be somewhat similar as before.

Updating only the approximation method but setting an (unrealistic) forward temperature of 55C (the current default) results in significantly higher central heat pump COPs and higher heat pump investments.


<img width="888" alt="image" src="https://github.com/user-attachments/assets/b4b4f45a-d975-4a14-a6dc-ffac32b776e5">
<img width="885" alt="image" src="https://github.com/user-attachments/assets/0e42cc6a-2788-421f-ad6e-d7952bdeec2d">
<img width="756" alt="image" src="https://github.com/user-attachments/assets/0f1076ca-42d6-439c-a7a3-adf5475887e0">

[jupyter_notebook.txt](https://github.com/user-attachments/files/16378347/jupyter_notebook.txt)

[config.txt](https://github.com/user-attachments/files/16378356/config.txt)

[scenarios.txt](https://github.com/user-attachments/files/16378358/scenarios.txt)

fneum commented on doc/configtables/sector.csv:
```diff
@@ -5,9 +5,17 @@ biomass,--,"{true, false}",Flag to include biomass sector.
--- potential,--,float,maximum fraction of urban demand which can be supplied by district heating. Ignored where below current fraction.
--- progress,--,Dictionary with planning horizons as keys., Increase of today's district heating demand to potential maximum district heating share. Progress = 0 means today's district heating share. Progress = 1 means maximum fraction of urban demand is supplied by district heating
--- district_heating_loss,--,float,Share increase in district heat demand in urban central due to heat losses
+#NAME?,--,float,maximum fraction of urban demand which can be supplied by district heating. Ignored where below current fraction.
```
@amosschle reviewing now, something went wrong with the configtables (`#Name?`). Could you revert this?

amosschle replied:
My bad, thanks for spotting thiis. Should be fixed now.

fneum commented on config/config.default.yaml:
```diff
@@ -408,6 +408,15 @@ sector:
       2045: 0.8
       2050: 1.0
     district_heating_loss: 0.15
+    # check these numbers!
```
```suggestion
```
to be removed for final version

fneum commented on doc/release_notes.rst:
```diff
@@ -10,6 +10,8 @@ Release Notes
 Upcoming Release
 ================
 
+* Changed heat pump COP approximation for central heating to be based on Jensen et al. 2018 and a default forward temperature of 90C. This is more realistic for district heating than the previously used approximation method.
```
perhaps add link to paper.

fneum commented on doc/configtables/sector.csv:
```diff
@@ -6,6 +6,14 @@ industry,--,"{true, false}",Flag to include industry sector.
+-- return_temperature,°C,float,Return temperature in district heating. Must be lower than forward temperature
+-- heat_source_cooling,K,float,Cooling of heat source for heat pumps
+-- heat_pump_cop_approximation,,,
+-- refrigerant,--,"{ammonia, isobutane}",Heat pump refrigerant assumed for COP approximation
+-- heat_exchanger_pinch_point_temperature_difference,K,float,Heat pump pinch point temperature difference in heat exchangers assumed for approximation.
+-- isentropic_compressor_efficiency,--,float,Isentropic efficiency of heat pump compressor assumed for approximation. Must be between 0 and 1.
+-- heat_loss,--,float,Heat pump heat loss assumed for approximation. Must be between 0 and 1.
```
```suggestion
-- -- refrigerant,--,"{ammonia, isobutane}",Heat pump refrigerant assumed for COP approximation
-- -- heat_exchanger_pinch_point_temperature_difference,K,float,Heat pump pinch point temperature difference in heat exchangers assumed for approximation.
-- -- isentropic_compressor_efficiency,--,float,Isentropic efficiency of heat pump compressor assumed for approximation. Must be between 0 and 1.
-- -- heat_loss,--,float,Heat pump heat loss assumed for approximation. Must be between 0 and 1.
```
to mark subgroup of config entries

fneum commented on scripts/prepare_sector_network.py:
```diff
@@ -1826,10 +1826,18 @@ def add_heat(n, costs):
 
     cop = {
-        "air": xr.open_dataarray(snakemake.input.cop_air_total)
+        "air decentral": xr.open_dataarray(snakemake.input.cop_air_decentral_heating)
         .to_pandas()
         .reindex(index=n.snapshots),
-        "ground": xr.open_dataarray(snakemake.input.cop_soil_total)
+        "ground decentral": xr.open_dataarray(
+            snakemake.input.cop_soil_decentral_heating
+        )
+        .to_pandas()
+        .reindex(index=n.snapshots),
+        "air central": xr.open_dataarray(snakemake.input.cop_air_central_heating)
+        .to_pandas()
+        .reindex(index=n.snapshots),
+        "ground central": xr.open_dataarray(snakemake.input.cop_soil_central_heating)
         .to_pandas()
         .reindex(index=n.snapshots),
     }
```
just a thought, but maybe one can build a dictionary with filenames first and then do:

```py
cop = {k: xr.open_dataarray(fn).to_pandas().reindex(index=n.snapshots) for k, vn in cop.items()}
```

which would reduce code length

amosschle replied:
I agree. We could replace `soil` with `ground` everywhere (or vice versa), define something like `heat_pump_sources = ["air", "ground"]`, `heat_system_type=["central", "decentral"]`, maybe in `config.default`, and use dict comprehensions over `heat_pump_sources` & `heat_system_type` wherever possible. What do you think?

fneum replied:
Yes! I would vote for keeping `ground` and replacing `soil` since it changes less in the config files users are exposed to.

fneum commented on scripts/add_existing_baseyear.py:
```diff
@@ -639,29 +638,43 @@ def add_heating_capacities_installed_before_baseyear(
+            n=n,
+            baseyear=baseyear,
+            grouping_years=grouping_years_heat,
+            cop={
+                "air decentral": xr.open_dataarray(
+                    snakemake.input.cop_air_decentral_heating
+                )
+                .to_pandas()
+                .reindex(index=n.snapshots),
+                "ground decentral": xr.open_dataarray(
+                    snakemake.input.cop_soil_decentral_heating
+                )
+                .to_pandas()
+                .reindex(index=n.snapshots),
+                "air central": xr.open_dataarray(
+                    snakemake.input.cop_air_central_heating
+                )
+                .to_pandas()
+                .reindex(index=n.snapshots),
+                "ground central": xr.open_dataarray(
+                    snakemake.input.cop_soil_central_heating
+                )
+                .to_pandas()
+                .reindex(index=n.snapshots),
+            },
```
see other similar comment to make this more compact through list comprehension

amosschle replied:
This should be fixed now with the new structure using only one file for COPs

fneum commented on scripts/add_existing_baseyear.py:
```diff
@@ -639,29 +638,43 @@ def add_heating_capacities_installed_before_baseyear(
-            gshp_cop,
-            time_dep_hp_cop,
-            costs,
-            default_lifetime,
+            n=n,
```
style suggestion: to make code less nested, perhaps created variables `cop` and `existing_heating` before function call and just pass object?

amosschle replied:
Hm, can do but that would create unnecessary memory load, wouldn't it?

fneum replied:
Also fine to leave as is, but I don't think it would create a memory overhead since no copy is made.

fneum commented on scripts/build_cop_profiles/BaseCopApproximator.py:
```diff
@@ -0,0 +1,82 @@
+        """
+        pass
+
+    def celsius_to_kelvin(
```
Is it necessary to duplicate the functions, once with `@staticmethod` decorator?

amosschle replied:
It very much isn't.

fneum reviewed (changes requested):
Looks mostly good to me. I like the structure!

One thing I am wondering about is whether this new COP calculation method should rather be integrated into `atlite` (i.e. here: https://github.com/PyPSA/atlite/blob/16c02fd4aa1816f73d96ed5ae52b1be69f671b18/atlite/convert.py#L259-L323). 

Bit unfortunate that I missed to tell you about this feature of `atlite` earlier. I can understand if you don't want to reimplement this now, but it would have some advantages, mainly that it would be available outside PyPSA-Eur.

amosschle commented:
> Looks mostly good to me. I like the structure, and it's very well documented!
> 
> One thing I am wondering about is whether this new COP calculation method should rather be integrated into `atlite` (i.e. here: https://github.com/PyPSA/atlite/blob/16c02fd4aa1816f73d96ed5ae52b1be69f671b18/atlite/convert.py#L259-L323).
> 
> Bit unfortunate that I missed to tell you about this feature of `atlite` earlier. I can understand if you don't want to reimplement this now, but it would have some advantages, mainly that it would be available outside PyPSA-Eur.

I agree. I'd suggest to merge it to `PyPSA-Eur` and move it to `atlite` later. Happy to open an issue for that.

fneum commented:
> I agree. I'd suggest to merge it to PyPSA-Eur and move it to atlite later. Happy to open an issue for that.

Yes, let's do that. Let me know when it's ready to merge. Read the Docs builds I will fix separately.

amosschle commented:
@fneum Thanks for the review! This is good to go from my side.

amosschle commented:
Ready for re-review. 
- I've made some structural changes that beautify the handling of different kinds of heat systems, system types etc. as enums and moved that to a `definitions` module, as suggested by @lkstrp. Hope that's fine.
- I'm getting slightly different results compared to a few days ago but I'm seeing that there were some changes in `master`, which I've merged into the feature that might be the reason (e.g. the clustering changed and COPs as a consequence)
- I've also adjusted the scripts for perfect foresight and myopic mode.
- Docs are not building but the error message is a little cryptic. Maybe @fneum, you can help?

lisazeyen commented:
> Ready for re-review.
> 
> * I've made some structural changes that beautify the handling of different kinds of heat systems, system types etc. as enums and moved that to a `definitions` module, as suggested by @lkstrp. Hope that's fine.
> * I'm getting slightly different results compared to a few days ago but I'm seeing that there were some changes in `master`, which I've merged into the feature that might be the reason (e.g. the clustering changed and COPs as a consequence)
> * I've also adjusted the scripts for perfect foresight and myopic mode.
> * Docs are not building but the error message is a little cryptic. Maybe @fneum, you can help?

- the Docs were not building through a changed syntax for the intersphinx mapping, which unrelated to changes in your PR, should be fixed in the master branch now

lisazeyen commented:
@amosschle really nice, code looks super clean and well documented! We should use the same approach for other sectors as well.

amosschle commented:
**Note for future reference:**
- #857 suggests using Lorenz COPs. [Pieper et al. 2020](https://www.sciencedirect.com/science/article/abs/pii/S0360544220311014) suggests an approximation via Jensen et al. (implemented in this PR) to be superior (the authors test both and other methods against a thermodynamic simulation model for a Danish district heating system
- This PR does not touch household heat pump COPs. This is still approximated via [Staffell et al. 2012](https://pubs.rsc.org/en/content/articlelanding/2012/ee/c2ee22653g/unauth). An alternative would be using values/method from the [When2Heat dataset](https://www.nature.com/articles/s41597-019-0199-y). My preliminary analysis (to be investigated further) suggests somewhat similar COP values.

fneum commented on scripts/add_existing_baseyear.py:
```diff
@@ -24,6 +24,10 @@
 from add_electricity import sanitize_carriers
 from prepare_sector_network import cluster_heat_buses, define_spatial, prepare_costs
 
+from scripts.definitions.heat_sector import HeatSector
+from scripts.definitions.heat_system import HeatSystem
+from scripts.definitions.heat_system_type import HeatSystemType
```
small thing, but I think one could just `from definitions.heat_sector import HeatSector`

amosschle replied:
Thanks for spotting. Happy to change it in #1180
